### PR TITLE
Initial value for persistent ControlObjects

### PR DESCRIPTION
**Pull request by Be-ing**
This lets constructors specify an initial value for persistent ControlObjects when there is not a valid value found in mixxx.cfg. I have added support for this to ControlPushButton. Should I add it for other ControlObject subclasses too? If so, which ones?

The first planned use of this is for the EffectUnit deck assignment switches in #1103.

### CODE REVIEW

File: src/control/control.cpp [hunk 1]
```diff
@@ -42,16 +42,19 @@ ControlDoublePrivate::ControlDoublePrivate(ConfigKey key,
...
-            value = pConfig->getValueString(m_key).toDouble();
+        if (pConfig != nullptr) {
+            bool conversionWorked;
+            value = pConfig->getValueString(m_key).toDouble(&conversionWorked);
+            if (!conversionWorked) {
+                value = initialValue;
+            }
         }
     }
     m_defaultValue.setValue(0);
```
Comment (rryan) [inline src/control/control.cpp]: I was thinking this would be the equivalent of calling "setDefaultValue" so it would replace the 0 here.

File: src/control/control.cpp [hunk 2]
```diff
@@ -42,16 +42,19 @@ ControlDoublePrivate::ControlDoublePrivate(ConfigKey key,
...
-void ControlDoublePrivate::initialize() {
+void ControlDoublePrivate::initialize(double initialValue) {
     double value = 0;
     if (m_bPersistInConfiguration) {
         UserSettingsPointer pConfig = ControlDoublePrivate::s_pUserConfig;
-        if (pConfig != NULL) {
-            // Assume toDouble() returns 0 if conversion fails.
-            value = pConfig->getValueString(m_key).toDouble();
+        if (pConfig != nullptr) {
+            bool conversionWorked;
```
Comment (rryan) [inline src/control/control.cpp]: ConfigObject now has a getValue method that does the type conversions so you can do:

```c++
value = defaultValue;
if (m_bPersistInConfiguration) { 
  UserSettingsPointer pConfig = ControlDoublePrivate::s_pUserConfig;
  if (pConfig != nullptr) { 
    value = m_pConfig->getValue(m_key, value);
  }
}
```

**Review by rryan — CHANGES_REQUESTED** [review]
Nice -- I think it's fine to leave adding the param to other sub-classes until someone needs it.

File: src/control/control.cpp (hunk 2 above)
Comment (rryan) [inline src/control/control.cpp]: Oh sorry, it's not there yet for doubles. @daschuer is adding it in another PR.

File: src/control/control.cpp (hunk 1 above)
Comment (rryan) [inline src/control/control.cpp]: (and for clarity it would be nice to have the names of all the initialValue vars be defaultValue)

**Comment by rryan** [comment]
Could you add a test to src/test/controlobjecttest.cpp?

File: src/control/control.cpp (hunk 2 above)
Comment (Be-ing) [inline src/control/control.cpp]: Regardless, there would need to be a way to differentiate between no valid value provided and 0.

File: src/control/control.cpp (hunk 1 above)
Comment (Be-ing) [inline src/control/control.cpp]: Yeah, that's what I initially did but I wasn't sure if that would always be desirable. I suppose it would be easy enough to override with another call to setDefaultValue() though.

File: src/control/control.cpp (hunk 2 above)
Comment (rryan) [inline src/control/control.cpp]: Right, this would do that -- the second argument to getValue is the default value to return if m_key is not present or malformatted. It would be equivalent to what you have here.

Comment (rryan) [inline src/control/control.cpp]: Added in 3901b751888265c831ba692392aed171797b917c.

File: src/test/effectslottest.cpp [hunk 3]
```diff
@@ -57,13 +57,16 @@ TEST_F(EffectSlotTest, ControlsReflectSlotState) {
     // Check the controls reflect the state of their loaded effect.
     EffectPointer pEffect = m_pEffectsManager->instantiateEffect(manifest.id());
 
-    // Enabled defaults to true in both effects and the slot.
-    pEffect->setEnabled(false);
-    EXPECT_DOUBLE_EQ(1.0, ControlObject::get(ConfigKey(group, "enabled")));
+    // Enabled defaults to false in both effects and the slot.
+    EXPECT_FALSE(pEffect->enabled());
```
Comment (rryan) [inline src/test/effectslottest.cpp]: These will fail before #1103 is in right?

**Review by rryan — APPROVED** [review]

File: src/test/effectslottest.cpp (hunk 3 above)
Comment (Be-ing) [inline src/test/effectslottest.cpp]: Whoops, didn't mean to commit that.

Comment (Be-ing) [inline src/test/effectslottest.cpp]: Force pushed revised commit without changes to this file.

File: src/test/controlobjecttest.cpp [hunk 4]
```diff
@@ -49,4 +49,16 @@ TEST_F(ControlObjectTest, aliasRetrieval) {
     EXPECT_EQ(ControlObject::getControl(ckAlias), co.get());
 }
 
+TEST_F(ControlObjectTest, persistence) {
```
Comment (rryan) [inline src/test/controlobjecttest.cpp]: Thanks for the test! To emulate shutting down Mixxx you could destroy the user settings in between creations of the CO:

```c++
// create CO with persist and check its value
m_pConfig->save();
m_pConfig.reset(new UserSettings(getTestDataDir().filePath("test.cfg")));
ControlDoublePrivate::setUserConfig(m_pConfig);
// create new CO with persist and check its value
```

You could also test other cases (invalid value in the config) by setting it manually to an invalid or empty value:
```c++
m_pConfig->set(ck, "notanumber");
```

File: src/test/controlobjecttest.cpp [hunk 5]
```diff
@@ -56,9 +56,29 @@ TEST_F(ControlObjectTest, persistence) {
...
+    // a user editing the stored CO value to an invalid value
+    delete testCo2;
+    m_pConfig->set(ck, QString("NotANumber"));
+    m_pConfig->save();
+    m_pConfig.clear();
+    m_pConfig = UserSettingsPointer(
+                    new UserSettings(getTestDataDir().filePath("test.cfg")));
+    ControlDoublePrivate::setUserConfig(m_pConfig);
+    ControlObject* testCo3 = new ControlObject(ck, true, false, true, 3.0);
+    EXPECT_DOUBLE_EQ(3.0, testCo3->get());
```
Comment (rryan) [inline src/test/controlobjecttest.cpp]: ```delete testCo3;``` here 
you can use std::unique_ptr to avoid having to manually manage the memory (see std::make_unique above)
